Detectors/TRD/reconstruction/src/CruRawReader.cxx

```diff
@@ -198,214 +160,113 @@ bool CruRawReader::compareRDH(const o2::header::RDHAny* firstrdh, const o2::head
+int CruRawReader::processHBFs()
 {
-  if (mHeaderVerbose) {
-    LOG(info) << "PROCESS HBF starting at " << std::hex << (void*)mDataPointer << " already read in : " << datasizealreadyread;
-  }
-  mDataRDH = reinterpret_cast<const o2::header::RDHAny*>(mDataPointer);
-  mOpenRDH = reinterpret_cast<const o2::header::RDHAny*>((const char*)mDataPointer);
-  auto rdh = mDataRDH;
-  uint32_t totaldataread = 0;
-  if (mHeaderVerbose) {
-    LOG(info) << " mem : " << o2::raw::RDHUtils::getMemorySize(rdh) << " headersize : " << o2::raw::RDHUtils::getHeaderSize(rdh);
-  }
-  mState = CRUStateHalfCRUHeader;
-  uint32_t currentsaveddatacount = 0;
+
+  const o2::header::RDHAny* rdh = reinterpret_cast<const o2::header::RDHAny*>(mCurrRdhPtr);
+  auto rdhPrevious = rdh;
+  bool firstRdh = true;
+  uint32_t totalDataInputSize = 0;
   mTotalHBFPayLoad = 0;
-  int loopcount = 0;
-  int counthalfcru = 0;
-  // store the endpoint and cruid for later checking of subsequent rdh for integrity
-  const o2::header::RDHAny* firstRDH;
-  firstRDH = reinterpret_cast<const o2::header::RDHAny*>(reinterpret_cast<const char*>(rdh));
-  mHBFoffset32 = 0;
+
   // loop until RDH stop header
   while (!o2::raw::RDHUtils::getStop(rdh)) { // carry on till the end of the event.
     if (mHeaderVerbose) {
-      LOG(info) << "----------------------------------------------";
-      LOG(info) << " rdh first word 0x" << std::hex << (uint32_t)*mDataPointer;
-      LOG(info) << " rdh first word is sitting at 0x" << std::hex << (void*)mDataPointer;
+      LOG(info) << "Current RDH is as follows:";
       o2::raw::RDHUtils::printRDH(rdh);
-      dumpRDHAndNextHeader(rdh);
     }
     if (!checkRDH(rdh)) {
-      if (mMaxErrsPrinted > 0) {
-        LOG(error) << "RDH validity failure : Call on call, an flp is not configured. Check the tracklets per halfchamber id for gaps of 30 or larger to figure out which one, feeid:" << std::hex << o2::raw::RDHUtils::getFEEID(rdh) << " cruid:" << o2::raw::RDHUtils::getCRUID(rdh) << " packetcounter:" << o2::raw::RDHUtils::getPacketCounter(rdh);
-        checkNoErr();
-      }
-      return false; // dump and run.
+      return -1;
     }
-    if (!compareRDH(firstRDH, rdh)) { // compare previous rdh detector info to the current rdh, they must be the same.
-      return false;                   // dump and run.
+    if (!firstRdh && !compareRDH(rdhPrevious, rdh)) {
+      // previous and current RDHs are inconsistent, this should not happen
+      return -1;
     }
-    firstRDH = rdh;
+    rdhPrevious = rdh;
+    firstRdh = false;
     auto headerSize = o2::raw::RDHUtils::getHeaderSize(rdh);
     auto memorySize = o2::raw::RDHUtils::getMemorySize(rdh);
-    if (memorySize == 0) {
-      LOG(warn) << "rdh memory size is zero";
-      return false; // get out of here if the rdh says it has nothing.
-    }
     auto offsetToNext = o2::raw::RDHUtils::getOffsetToNext(rdh);
     auto rdhpayload = memorySize - headerSize;
```

Comment:
Could memorySize < headerSize at any point? Then the memcpy gets problematic.

Reply:
No, because the memory size contains the header size. Only of the RDH itself would be corrupted this could not be true, but then the above `checkRDH()` should fail and the whole data for that RDH is dumped